File: algo-java/src/main/java/com/time1043/started004/SelectBubbleInsertSort.java
package com.time1043.started004;

// https://leetcode.com/problems/sort-an-array/
public class SelectBubbleInsertSort {
    public static void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public static void selectSort(int[] nums) {
        // Consider the boundary
        if (nums == null || nums.length < 2) return;

        // Calculate minIdx
        int minIdx;
        for (int i = 0; i < nums.length - 1; i++) {
            minIdx = i;
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[j] < nums[minIdx]) minIdx = j;
            }
            // swap(nums, i, minIdx);
            if (minIdx != i) swap(nums, i, minIdx);
        }
    }

    public static void bubbleSort(int[] nums) {
        // Consider the boundary
        if (nums == null || nums.length < 2) return;

        // By comparing neighboring positions, the larger one bubbles backward
        for (int i = nums.length - 1; i > 0; i--) {
            for (int j = 0; j < i; j++) {
                if (nums[j] > nums[j + 1]) swap(nums, j, j + 1);
            }
        }
    }

    public static void insertSort(int[] nums) {
        // Consider the boundary
        if (nums == null || nums.length < 2) return;

        // Ensure that 0~i is orderly, i=1~n-1
        for (int i = 1; i <= (nums.length - 1); i++) {
            for (int j = i; j > 0; j--) {
                if (nums[j] < nums[j - 1]) swap(nums, j, j - 1);
                else break;
            }
        }
    }
}

File: algo-kotlin/src/main/kotlin/section004sort/SelectBubbleInsertSort.kt
package com.time1043.section004sort

fun IntArray.swap(i: Int, j: Int) {
    val tmp = this[i]
    this[i] = this[j]
    this[j] = tmp
}

class SelectBubbleInsertSort {
    fun selectionSort(nums: IntArray) {
        // Consider the boundary
        if (nums.isEmpty() || nums.size < 2) return

        // Calculate minIdx
        var minIdx = 0
        for (i in 0 until nums.size - 1) {
            minIdx = i
            for (j in i + 1 until nums.size) {
                if (nums[j] < nums[minIdx]) minIdx = j
            }
            nums.swap(i, minIdx)
        }
    }

    fun bubbleSort(nums: IntArray) {
        // Consider the boundary
        if (nums.isEmpty() || nums.size < 2) return

        // By comparing neighboring positions, the larger one bubbles backward
        for (i in nums.lastIndex downTo 1) {
            for (j in 0 until i) {
                if (nums[j] > nums[j + 1]) nums.swap(j, j + 1)
            }
        }
    }

    fun insertSort(nums: IntArray) {
        // Consider the boundary
        if (nums.isEmpty() || nums.size < 2) return

        // Ensure that 0~i is orderly, i=1~n-1
        for (i in 1..(nums.lastIndex)) {
            for (j in i downTo 1) {
                if (nums[j] < nums[j - 1]) nums.swap(j, j - 1)
            }
        }
    }
}